src/viam/sdk/common/proto_type.hpp

```diff
@@ -27,64 +27,82 @@ struct get_helper {
-    explicit ProtoType(double d) : proto_type_(std::move(d)) {}
-    explicit ProtoType(AttributeMap m) : proto_type_(std::move(m)) {}
-    explicit ProtoType(std::vector<std::shared_ptr<ProtoType>> v) : proto_type_(std::move(v)) {}
-    explicit ProtoType(const google::protobuf::Value& value);
-
-    google::protobuf::Value proto_value();
-    friend bool operator==(const ProtoType& lhs, const ProtoType& rhs);
-
-    template <typename T>
-    auto get() {
-        return prototype_details::get_helper<T>::get(proto_type_);
-    }
-
-    template <typename T>
-    auto get() const {
-        return prototype_details::get_helper<T>::get(proto_type_);
-    }
+    class value_type;
+
+    using map_type = std::unordered_map<std::string, value_type>;
+
+    class value_type {
+       public:
+        template <typename T>
+        auto get() {
+            return value_type_details::get_helper<T>::get(value_);
+        }
+
+        template <typename T>
+        auto get() const {
+            return value_type_details::get_helper<T>::get(value_);
+        }
+
+        using base_value_types_ =
+            boost::variant<boost::blank, std::string, const char*, int, float, double, bool>;
+        using value_types = boost::make_recursive_variant<
+            base_value_types_,
+            std::unordered_map<std::string, boost::recursive_variant_>,
+            std::vector<boost::recursive_variant_>>::type;
+
+        value_type();
+        value_type(std::string s);
+        value_type(const char* c);
+        value_type(int i);
+        value_type(float f);
+        value_type(double d);
+        value_type(bool b);
+        value_type(std::vector<value_type> v);
+        value_type(map_type);
+        value_type(google::protobuf::Value value);
+
+        google::protobuf::Value to_proto() const;
+
+       private:
+        value_types value_;
+    };
+
+    AttributeMap();
+    ~AttributeMap();
+    // CR erodkin: add comments to all of this probably
+    google::protobuf::Struct to_proto() const;
+
+    bool contains(const std::string& key) const;
+
+    // CR erodkin: consider how to return a ref or pointer to the value type such that we don't have
+    // to copy the `value_type`, while retaining guarantees of const-ness.
+    value_type get(const std::string& key) const;
+
+    // set a new k/v pair in the AttributeMap
+    void set(std::string key, value_type::value_types value) const;
+    void set(std::string key, value_type value) const;
+
+    // append a single value to a list type within the AttributeMap
+    void append(const std::string& key, value_type::value_types value);
+    void append(const std::string& key, value_type value);
+
+    // add a single key/value pair to a map type within the AttributeMap
+    void insert(const std::string& key, std::pair<std::string, value_type>);
+    void insert(const std::string& key, std::pair<std::string, value_type::value_types>);
```

Comment:
It may just be better to expose the `map_` rather than try to define your own wrapper above the unordered_map interface.

Reply:
Yeah, I agree this is trying to do to much. I'm wondering though if it's better to keep the `AttributeMap` class around and expose `map_` or to just define `value_type` at the top level and have `AttributeMap` just be a typedef on the map type. The latter seems more intuitive to me personally, though it does perhaps make `get` helpers slightly more awkward if they're just free standing functions rather than a class method.